docs/dev/topology-over-raft.md

```diff
@@ -216,3 +216,94 @@ There are also a few static columns for cluster-global properties:
 - `current_cdc_generation_uuid` - the UUID of the last introduced CDC generation (used to access its data)
 - `global_topology_request` - if set, contains one of the supported global topology requests
 - `new_cdc_generation_data_uuid` - used in `commit_cdc_generation` state, the UUID of the generation to be committed
+
+# Join procedure
+
+In topology on raft mode, new nodes need to go through a new handshake procedure
+before they can join the cluster, including joining group 0. The handshake
+happens during raft discovery and replaced the old GROUP0_MODIFY_CONFIG.
+
+Two RPC verbs are introduced:
+
+- `JOIN_NODE_REQUEST`
+  Sent by the node that wishes to join the cluster. It contains some basic
+  information about the node that will have to be verified by the coordinator.
+
+- `JOIN_NODE_REPLY`
+  Sent by the topology coordinator back to the joining node. It contains
+  coordinator's decision about the node - whether the request was accepted
+  or rejected.
+
+In order to correlate requests with replies, the joining node uses a UUID called
+"join cookie". Until the joining node acknowledges a reply from the coordinator,
```

Comment:
I did not look at the code how this "join cookie" is used, but can host_id can be used instead?

Reply:
I'm guessing that host_id is generated and persisted early in the procedure. If the procedure crashes after persisting host_id, the next attempt might use some different parameters (which ones though?) but a different host_id.

Reply:
If "host_id is generated and persisted early in the procedure" then after the crash the same host_id will be used. Theoretically one may change some parameters (cluster name? Snitch?) and and restart, but if we persist those before start joining we can detect the change and fail early. If features are changed we can detect it in join_request rpc since whatever was added into the topology will not match.

Reply:
One way to deal with lack of good names for the verb is to rename group0_modify_config to group0_modify_config_legacy and keep using the name group0_modify_config

Reply:
> One way to deal with lack of good names for the verb is to rename group0_modify_config to group0_modify_config_legacy and keep using the name group0_modify_config

I suppose that it's not a problem that the `join_node` name is less clear than `group0_modify_config`; the bigger problem is that we have two RPCs now which are both used in the handshake named `join_node_request` and `join_node_response`, and _both_ of them are two-way. It's confusing to talk about "response to join_node_request" etc.